### PR TITLE
remove serializing WalkDir() across all buckets/prefixes on SSDs

### PR DESCRIPTION
**Pull request by harshavardhana**
## Description
remove serializing WalkDir() across all buckets/prefixes

## Motivation and Context
slower drives get knocked off they are too slow via active 
monitoring, we do not need to block calls arbitrarily.

Serializing adds latencies for already slow calls, remove 
it and we can revisit this if needed in the future.

Also add selection with context when writing to `out <-` 
channel, to avoid any potential blocks.

## Types of changes
- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Optimization (provides speedup with no functional changes)
- [x] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:
- [ ] Fixes a regression (If yes, please add `commit-id` or `PR #` here)
- [ ] Unit tests added/updated
- [ ] Internal documentation updated
- [ ] Create a documentation update request [here](https://github.com/minio/docs/issues/new?label=doc-change,title=Doc+Updated+Needed+For+PR+github.com%2fminio%2fminio%2fpull%2fNNNNN)

### CODE REVIEW

**Comment by harshavardhana** [comment]
Never seen a race like this

```
==================
==================
WARNING: DATA RACE
Write at 0x00c005b9c038 by goroutine 273192:
  runtime.racewriterange()
      <autogenerated>:1 +0x29
  internal/poll.ignoringEINTRIO()
      /opt/hostedtoolcache/go/1.20.6/x64/src/internal/poll/fd_unix.go:794 +0x439
  internal/poll.(*FD).Read()
      /opt/hostedtoolcache/go/1.20.6/x64/src/internal/poll/fd_unix.go:163 +0x2e
  os.(*File).read()
      /opt/hostedtoolcache/go/1.20.6/x64/src/os/file_posix.go:31 +0xc7
  os.(*File).Read()
      /opt/hostedtoolcache/go/1.20.6/x64/src/os/file.go:118 +0x98
  io.ReadAtLeast()
      /opt/hostedtoolcache/go/1.20.6/x64/src/io/io.go:332 +0xdd
  io.ReadFull()
      /opt/hostedtoolcache/go/1.20.6/x64/src/io/io.go:351 +0xee
  github.com/minio/minio/cmd.readXLMetaNoData()
      /home/runner/work/minio/minio/cmd/xl-storage-format-v2.go:697 +0xd3
  github.com/minio/minio/cmd.(*xlStorage).readMetadataWithDMTime()
      /home/runner/work/minio/minio/cmd/xl-storage.go:425 +0x504
  github.com/minio/minio/cmd.(*xlStorage).readMetadata()
      /home/runner/work/minio/minio/cmd/xl-storage.go:433 +0x1a4a
  github.com/minio/minio/cmd.(*xlStorage).WalkDir.func3()
      /home/runner/work/minio/minio/cmd/metacache-walk.go:307 +0x1998
  github.com/minio/minio/cmd.(*xlStorage).WalkDir.func3()
      /home/runner/work/minio/minio/cmd/metacache-walk.go:355 +0x2244
  github.com/minio/minio/cmd.(*xlStorage).WalkDir.func3()
      /home/runner/work/minio/minio/cmd/metacache-walk.go:355 +0x2244
  github.com/minio/minio/cmd.(*xlStorage).WalkDir.func3()
      /home/runner/work/minio/minio/cmd/metacache-walk.go:355 +0x2244
  github.com/minio/minio/cmd.(*xlStorage).WalkDir.func3()
      /home/runner/work/minio/minio/cmd/metacache-walk.go:355 +0x2244
  github.com/minio/minio/cmd.(*xlStorage).WalkDir.func3()
      /home/runner/work/minio/minio/cmd/metacache-walk.go:355 +0x2244
  github.com/minio/minio/cmd.(*xlStorage).WalkDir.func3()
      /home/runner/work/minio/minio/cmd/metacache-walk.go:355 +0x2244
  github.com/minio/minio/cmd.(*xlStorage).WalkDir()
      /home/runner/work/minio/minio/cmd/metacache-walk.go:363 +0x909
  github.com/minio/minio/cmd.(*xlStorageDiskIDCheck).WalkDir()
      /home/runner/work/minio/minio/cmd/metacache-walk.go:373 +0x2f7
  github.com/minio/minio/cmd.listPathRaw.func3()
      /home/runner/work/minio/minio/cmd/metacache-set.go:933 +0x252

Previous read at 0x00c005b9c038 by goroutine 273178:
  encoding/binary.bigEndian.Uint16()
      /opt/hostedtoolcache/go/1.20.6/x64/src/encoding/binary/binary.go:139 +0x1b6
  github.com/tinylib/msgp/msgp.readBytesBytes()
      /home/runner/go/pkg/mod/github.com/tinylib/msgp@v1.1.8/msgp/read_bytes.go:815 +0x16a
  github.com/tinylib/msgp/msgp.ReadBytesZC()
      /home/runner/go/pkg/mod/github.com/tinylib/msgp@v1.1.8/msgp/read_bytes.go:862 +0xfc
  github.com/minio/minio/cmd.decodeVersions()
      /home/runner/work/minio/minio/cmd/xl-storage-format-v2.go:818 +0xe1
  github.com/minio/minio/cmd.xlMetaBuf.AllHidden()
      /home/runner/work/minio/minio/cmd/xl-storage-format-v2.go:2147 +0xec
  github.com/minio/minio/cmd.(*xlStorage).WalkDir.func1()
      /home/runner/work/minio/minio/cmd/metacache-walk.go:97 +0x6e
  github.com/minio/minio/cmd.(*xlStorage).WalkDir.func3()
      /home/runner/work/minio/minio/cmd/metacache-walk.go:318 +0x2010
  github.com/minio/minio/cmd.(*xlStorage).WalkDir()
      /home/runner/work/minio/minio/cmd/metacache-walk.go:363 +0x909
  github.com/minio/minio/cmd.(*xlStorageDiskIDCheck).WalkDir()
      /home/runner/work/minio/minio/cmd/metacache-walk.go:373 +0x2f7
  github.com/minio/minio/cmd.listPathRaw.func3()
      /home/runner/work/minio/minio/cmd/metacache-set.go:933 +0x252
```

Restarting the tests

**Comment by harshavardhana** [comment]
@klauspost ^^

File: cmd/metacache-walk.go
Comment (klauspost) [inline cmd/metacache-walk.go:242]: Seems like you are creating the race here, by referencing `meta.metadata` after sending it (you do this below as well). There is no reason for that change.

**Review by klauspost — COMMENTED** [review]
I am plainly against this change. You will just end up creating a scenario where listing will kill all operations on servers instead of just progressively slowing down listings.

This is a BAD idea, and I will not approve it. You are welcome to ignore my objections, but you will see loads of clogged up servers.

If you implement this as an SSD opt-in, and maybe enable where you can *reliably* detect SSD, I don't have any problems. But as a general solution this is poison.

File: cmd/metacache-walk.go
Comment (harshavardhana) [reply cmd/metacache-walk.go:242]: I thought we need to do this after we wrote to channel, looks like it's not possible.

Let me revert it.

**Comment by harshavardhana** [comment]
> I am plainly against this change. You will just end up creating a scenario where listing will kill all operations on servers instead of just progressively slowing down listings.
> 
> This is a BAD idea, and I will not approve it. You are welcome to ignore my objections, but you will see loads of clogged up servers.
> 
> If you implement this as an SSD opt-in, and maybe enable where you can _reliably_ detect SSD, I don't have any problems. But as a general solution this is poison.

Yes @klauspost I totally agree, let me bring an environment variable that changes the default behavior.

**Comment by harshavardhana** [comment]
> I am plainly against this change. You will just end up creating a scenario where listing will kill all operations on servers instead of just progressively slowing down listings.
> 
> This is a BAD idea, and I will not approve it. You are welcome to ignore my objections, but you will see loads of clogged up servers.
> 
> If you implement this as an SSD opt-in, and maybe enable where you can _reliably_ detect SSD, I don't have any problems. But as a general solution this is poison.

Added detection for HDDs v/s SSDs

File: cmd/xl-storage.go
Comment (klauspost) [inline cmd/xl-storage.go]:
```suggestion
	if info.Rotational == nil || *info.Rotational {
```

Keeps current for all non-linux.

**Review by klauspost — COMMENTED** [review]
LGTM, when keeping current, if undetectable.

**Review by klauspost — APPROVED** [review]
LGTM

**Review by vadmeste — COMMENTED** [review]
LGTM

**Review by vadmeste — APPROVED** [review]
LGTM

**Pull request merged**
